<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Example MR run</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">UKB IDP Genetic Architecture</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/ukb_idp_genetic_arch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Example MR run</h1>
<h4 class="date">19 November, 2020</h4>

</div>


<pre class="r"><code>rm(list = ls())
library(ggplot2)
theme_set(theme_bw(base_size = 15))
source(&#39;https://gist.githubusercontent.com/liangyy/43912b3ecab5d10c89f9d4b2669871c9/raw/3ca651cfa53ffccb8422f432561138a46e93710f/my_ggplot_theme.R&#39;)
source(&#39;https://raw.githubusercontent.com/liangyy/misc-tools/master/plot_tool/plot_tools.R&#39;)
library(dplyr)
options(stringsAsFactors = F)
library(patchwork)
library(data.table)
options(datatable.fread.datatable = F)
library(TwoSampleMR)
source(&#39;rlib.R&#39;)</code></pre>
<div id="about" class="section level1">
<h1><span class="header-section-number">1</span> About</h1>
<p>As the downstream analysis of ImageXcan run. We want to perform Mendelian Randomization to see if the IDP being identified is the exposure of the outcome. Here I take an example <code>weekly_alcohol</code> and <code>IDP-25882</code> (as suggested by <a href="imagexcan_round_1_t1.html">this post</a>). I use the <code>TwoSampleMR</code> package to perform MR. Since I could not find <code>weekly_alcohol</code> in UKB, I will use <a href="https://biobank.ctsu.ox.ac.uk/crystal/field.cgi?id=1558">Alcohol intake frequency</a> <code>ukb-b-5779</code> in open gwas database <a href="https://gwas.mrcieu.ac.uk/datasets/ukb-b-5779/">link</a>.</p>
</div>
<div id="load-idp-gwas" class="section level1">
<h1><span class="header-section-number">2</span> Load IDP GWAS</h1>
<pre class="r"><code>idp_code = &#39;IDP-25882&#39;
idp_gwas = list()
for(i in 1 : 22) {
  idp_gwas[[length(idp_gwas) + 1]] = arrow::read_parquet(paste0(&#39;~/Desktop/tmp/ukb_idp/idp_gwas/trans_qtl.T1.chr&#39;, i, &#39;/&#39;, idp_code, &#39;.parquet&#39;))
}
idp_gwas = do.call(rbind, idp_gwas)
qqplot_by_group(idp_gwas$pval[ idp_gwas$pval &lt; 0.01 ], 1, pval_cutoff = 0.01) + th</code></pre>
<p><img src="example_mr_files/figure-html/unnamed-chunk-104-1.png" width="672" /></p>
<p>We need to load more meta information of the SNPs.</p>
<pre class="r"><code>snp_meta = list()
for(i in 1 : 22) {
  snp_meta[[length(snp_meta) + 1]] = read.table(paste0(&#39;~/Desktop/tmp/ukb_idp/subset_genotypes/IDP_HM3_finalPheno.chr&#39;, i, &#39;.bim&#39;), header = F) 
}
snp_meta = do.call(rbind, snp_meta)
snp_meta = snp_meta %&gt;% select(V2, V5, V6) %&gt;% rename(rsid = V2, ref = V5, alt = V6)
idp_gwas = left_join(idp_gwas, snp_meta, by = c(&#39;variant_id&#39; = &#39;rsid&#39;))</code></pre>
</div>
<div id="idp---alcohol-intake-frequency" class="section level1">
<h1><span class="header-section-number">3</span> IDP -&gt; alcohol intake frequency</h1>
<p>Set up the exposure GWAS.</p>
<pre class="r"><code>idp_exp_dat = format_data(
  data.frame(SNP = idp_gwas$variant_id, beta = idp_gwas$b, se = idp_gwas$b_se, effect_allele = idp_gwas$alt, other_allele = idp_gwas$ref)
)</code></pre>
<pre><code>## No phenotype name specified, defaulting to &#39;exposure&#39;.</code></pre>
<pre><code>## Warning in format_data(data.frame(SNP = idp_gwas$variant_id, beta = idp_gwas$b, : The following columns are not present but are helpful for harmonisation
## eaf</code></pre>
<pre><code>## Inferring p-values</code></pre>
<p>Do LD clumping on data.</p>
<pre class="r"><code>idp_exp_dat = clump_data(
  idp_exp_dat, 
  pop = &#39;EUR&#39;,
  clump_p1 = 5e-08,
  clump_p2 = 5e-08,
  clump_r2 = 0.001,
  clump_kb = 10000
)</code></pre>
<pre><code>## Please look at vignettes for options on running this locally if you need to run many instances of this command.</code></pre>
<pre><code>## Clumping HiYTOw, 1071649 variants, using EUR population reference</code></pre>
<pre><code>## Removing 1071639 of 1071649 variants due to LD with other variants or absence from LD reference panel</code></pre>
<p>Set up the outcome GWAS.</p>
<pre class="r"><code>ukb_code = &#39;ukb-b-5779&#39;
outcome_dat &lt;- extract_outcome_data(
    snps = idp_exp_dat$SNP,
    outcomes = ukb_code
)</code></pre>
<pre><code>## Extracting data for 10 SNP(s) from 1 GWAS(s)</code></pre>
<p>Run MR.</p>
<pre class="r"><code>if(!exists(&#39;dat_forward&#39;)) {
  dat_forward &lt;- harmonise_data(idp_exp_dat, outcome_dat)
}</code></pre>
<pre><code>## Harmonising exposure (HiYTOw) and Alcohol intake frequency. || id:ukb-b-5779 (ukb-b-5779)</code></pre>
<pre class="r"><code># Perform MR
res_forward &lt;- mr(dat_forward)</code></pre>
<pre><code>## Analysing &#39;HiYTOw&#39; on &#39;ukb-b-5779&#39;</code></pre>
<pre class="r"><code>res_forward %&gt;% pander::pander(caption = &#39;IDP -&gt; phenotype&#39;)</code></pre>
<table style="width:96%;">
<caption>IDP -&gt; phenotype (continued below)</caption>
<colgroup>
<col width="19%" />
<col width="18%" />
<col width="43%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">id.exposure</th>
<th align="center">id.outcome</th>
<th align="center">outcome</th>
<th align="center">exposure</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">HiYTOw</td>
<td align="center">ukb-b-5779</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
<td align="center">exposure</td>
</tr>
<tr class="even">
<td align="center">HiYTOw</td>
<td align="center">ukb-b-5779</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
<td align="center">exposure</td>
</tr>
<tr class="odd">
<td align="center">HiYTOw</td>
<td align="center">ukb-b-5779</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
<td align="center">exposure</td>
</tr>
<tr class="even">
<td align="center">HiYTOw</td>
<td align="center">ukb-b-5779</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
<td align="center">exposure</td>
</tr>
<tr class="odd">
<td align="center">HiYTOw</td>
<td align="center">ukb-b-5779</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
<td align="center">exposure</td>
</tr>
</tbody>
</table>
<table style="width:92%;">
<colgroup>
<col width="38%" />
<col width="9%" />
<col width="15%" />
<col width="13%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">method</th>
<th align="center">nsnp</th>
<th align="center">b</th>
<th align="center">se</th>
<th align="center">pval</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">MR Egger</td>
<td align="center">10</td>
<td align="center">-0.2375</td>
<td align="center">0.1308</td>
<td align="center">0.107</td>
</tr>
<tr class="even">
<td align="center">Weighted median</td>
<td align="center">10</td>
<td align="center">-0.08236</td>
<td align="center">0.03495</td>
<td align="center">0.01846</td>
</tr>
<tr class="odd">
<td align="center">Inverse variance weighted</td>
<td align="center">10</td>
<td align="center">-0.05078</td>
<td align="center">0.05682</td>
<td align="center">0.3715</td>
</tr>
<tr class="even">
<td align="center">Simple mode</td>
<td align="center">10</td>
<td align="center">-0.09031</td>
<td align="center">0.05115</td>
<td align="center">0.1112</td>
</tr>
<tr class="odd">
<td align="center">Weighted mode</td>
<td align="center">10</td>
<td align="center">-0.1081</td>
<td align="center">0.04239</td>
<td align="center">0.03119</td>
</tr>
</tbody>
</table>
<p>Visualization.</p>
<pre class="r"><code>dat_forward %&gt;% filter(pval.exposure &lt; 1e-6) %&gt;% ggplot() + 
  geom_vline(xintercept = 0, color = &#39;grey&#39;) + geom_hline(yintercept = 0, color = &#39;grey&#39;) + 
  geom_point(aes(x = beta.exposure, y = beta.outcome, size = -log10(pval.exposure))) + th </code></pre>
<p><img src="example_mr_files/figure-html/unnamed-chunk-110-1.png" width="672" /></p>
</div>
<div id="alcohol-intake-frequency---idp" class="section level1">
<h1><span class="header-section-number">4</span> alcohol intake frequency -&gt; IDP</h1>
<p>Load instrument GWAS.</p>
<pre class="r"><code>exp_dat2 &lt;- extract_instruments(outcomes = ukb_code)</code></pre>
<p>Load IDP as outcome GWAS.</p>
<pre class="r"><code>idp_dat = format_data(
  data.frame(SNP = idp_gwas$variant_id, beta = idp_gwas$b, se = idp_gwas$b_se, effect_allele = idp_gwas$alt, other_allele = idp_gwas$ref),
  type = &#39;outcome&#39;,
  snps = exp_dat2$SNP
)</code></pre>
<pre><code>## No phenotype name specified, defaulting to &#39;outcome&#39;.</code></pre>
<pre><code>## Warning in format_data(data.frame(SNP = idp_gwas$variant_id, beta = idp_gwas$b, : The following columns are not present but are helpful for harmonisation
## eaf</code></pre>
<pre><code>## Inferring p-values</code></pre>
<p>Run MR.</p>
<pre class="r"><code>if(!exists(&#39;dat_backward&#39;)) {
  dat_backward &lt;- harmonise_data(exp_dat2, idp_dat)
}</code></pre>
<pre><code>## Harmonising Alcohol intake frequency. || id:ukb-b-5779 (ukb-b-5779) and outcome (47t3ls)</code></pre>
<pre class="r"><code># Perform MR
res_backward &lt;- mr(dat_backward)</code></pre>
<pre><code>## Analysing &#39;ukb-b-5779&#39; on &#39;47t3ls&#39;</code></pre>
<pre class="r"><code>res_backward %&gt;% pander::pander(caption = &#39;phenotype -&gt; IDP&#39;)</code></pre>
<table style="width:94%;">
<caption>phenotype -&gt; IDP (continued below)</caption>
<colgroup>
<col width="19%" />
<col width="18%" />
<col width="13%" />
<col width="43%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">id.exposure</th>
<th align="center">id.outcome</th>
<th align="center">outcome</th>
<th align="center">exposure</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">ukb-b-5779</td>
<td align="center">47t3ls</td>
<td align="center">outcome</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
</tr>
<tr class="even">
<td align="center">ukb-b-5779</td>
<td align="center">47t3ls</td>
<td align="center">outcome</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
</tr>
<tr class="odd">
<td align="center">ukb-b-5779</td>
<td align="center">47t3ls</td>
<td align="center">outcome</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
</tr>
<tr class="even">
<td align="center">ukb-b-5779</td>
<td align="center">47t3ls</td>
<td align="center">outcome</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
</tr>
<tr class="odd">
<td align="center">ukb-b-5779</td>
<td align="center">47t3ls</td>
<td align="center">outcome</td>
<td align="center">Alcohol intake frequency. || id:ukb-b-5779</td>
</tr>
</tbody>
</table>
<table style="width:90%;">
<colgroup>
<col width="38%" />
<col width="9%" />
<col width="13%" />
<col width="13%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">method</th>
<th align="center">nsnp</th>
<th align="center">b</th>
<th align="center">se</th>
<th align="center">pval</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">MR Egger</td>
<td align="center">21</td>
<td align="center">-0.1001</td>
<td align="center">0.3481</td>
<td align="center">0.7768</td>
</tr>
<tr class="even">
<td align="center">Weighted median</td>
<td align="center">21</td>
<td align="center">0.049</td>
<td align="center">0.09797</td>
<td align="center">0.617</td>
</tr>
<tr class="odd">
<td align="center">Inverse variance weighted</td>
<td align="center">21</td>
<td align="center">0.05979</td>
<td align="center">0.08726</td>
<td align="center">0.4932</td>
</tr>
<tr class="even">
<td align="center">Simple mode</td>
<td align="center">21</td>
<td align="center">0.06854</td>
<td align="center">0.1636</td>
<td align="center">0.6797</td>
</tr>
<tr class="odd">
<td align="center">Weighted mode</td>
<td align="center">21</td>
<td align="center">0.04221</td>
<td align="center">0.1244</td>
<td align="center">0.7379</td>
</tr>
</tbody>
</table>
<p>Visualization.</p>
<pre class="r"><code>dat_backward %&gt;% ggplot() + 
  geom_vline(xintercept = 0, color = &#39;grey&#39;) + geom_hline(yintercept = 0, color = &#39;grey&#39;) + 
  geom_point(aes(x = beta.exposure, y = beta.outcome, size = -log10(pval.exposure))) + th </code></pre>
<p><img src="example_mr_files/figure-html/unnamed-chunk-114-1.png" width="672" /></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
